ksqldb-engine/src/test/java/io/confluent/ksql/query/PullQueryWriteStreamTest.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.query;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.pull.StreamedRowTranslator;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.KeyValueMetadata;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PullQueryWriteStreamTest {
+
+  private static final QueryId TEST_ID = new QueryId("test");
+  private static final int TEST_TIMEOUT_SECONDS = 60;
+  private static final int LIMIT = 5;
+
+  @Rule
+  public final Timeout timeout = Timeout.seconds(TEST_TIMEOUT_SECONDS);
+
+  @Mock private LogicalSchema schema;
+  private PullQueryWriteStream writeStream;
+  private ScheduledExecutorService executorService;
+
+  @Before
+  public void setUp() {
+    final StreamedRowTranslator translator = new StreamedRowTranslator(schema, Optional.empty());
+    writeStream = new PullQueryWriteStream(OptionalInt.of(LIMIT), translator);
+    writeStream.write(ImmutableList.of(StreamedRow.header(TEST_ID, schema)));
```

Comment:
Idly, does the header count against the limits?

Reply:
no, it doesn't (it actually doesn't get written to the queue it's just collected as metadata - something else writes it to the response, I don't know where exactly)